File: Python3/201 - 400/300 Longest Increasing Subsequence.py
# 56ms, 97.18%
class Solution:
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        res = []
        for n in nums:
            it = self.lower_bound(res, n)
            if it == -1:
                res.append(n)
            else:
                res[it] = n
        return len(res)

           
    def lower_bound(self, arr, x):
        if not arr or arr[-1] < x: return -1
        left, right = 0, len(arr) - 1
        while left < right:
            mid = (left + right) // 2
            if arr[mid] < x:
                left = mid + 1
            else:
                right = mid
        return left

File: 201 - 400/300 Longest Increasing Subsequence.cc
// 4ms, 96.80%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int lengthOfLIS(const vector<int> &nums) {
        vector<int> res;
        for (const auto n : nums) {
            auto it = lower_bound(res.begin(), res.end(), n);
            if (it == res.end()) {
                res.push_back(n);
            } else {
                *it = n;
            }
        }
        return res.size();
    }
};
